### PR TITLE
sql: require CONNECTION when creating SINK

### PR DESCRIPTION
**Pull request by sploiselle**
Require `KAFKA CONNECTION` on `SINK` creation

### Motivation

This PR adds a known-desirable feature. #13374

### Checklist

- [x] This PR has adequate test coverage / QA involvement has been duly considered.
- [ ] This PR evolves [an existing `$T ⇔ Proto$T` mapping](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/command-and-response-binary-encoding.md) (possibly in a backwards-incompatible way) and therefore is tagged with a `T-protobuf` label.
- [x] This PR includes the following [user-facing behavior changes](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/guide-changes.md#what-changes-require-a-release-note):

  - Require a `KAFKA CONNECTION` when creating a `SINK`.

### CODE REVIEW

**Comment by sploiselle** [comment]
Included in this PR was an attempt to start replacing some `SOURCE` tests with `CONNECTION` in lieu of `BROKER`, but I ran out of steam.

File: src/sql-parser/src/ast/defs/ddl.rs [hunk 1]
```diff
@@ -951,14 +951,13 @@ impl<T: AstInfo> AstDisplay for CreateSinkConnection<T> {
...
-                broker,
+                connection,
                 topic,
                 key,
                 consistency,
             } => {
-                f.write_str("KAFKA BROKER '");
-                f.write_node(&display::escape_single_quote_string(broker));
-                f.write_str("'");
+                f.write_str("KAFKA ");
```
Comment (guswynn) [inline src/sql-parser/src/ast/defs/ddl.rs:959]: `KAFKA CONNECTION`?

File: src/sql/src/plan/statement/ddl.rs [hunk 2]
```diff
@@ -1814,18 +1814,45 @@ pub fn plan_create_materialized_view(
...
+            // simply be a disjoint set of what `CONNECTION`s support.
+            for k in BTreeMap::<String, StringOrSecret>::from(connection.clone()).keys() {
+                if config_options.contains_key(k) {
+                    sql_bail!(
+                        "cannot set option {} for SINK using CONNECTION {}",
+                        k,
+                        scx.catalog.resolve_full_name(item.name())
+                    );
+                }
+            }
```
Comment (guswynn) [inline src/sql/src/plan/statement/ddl.rs:1850]: is there a new test for this? I'm having trouble wrapping my head around what its doing

File: src/sql/src/plan/statement/ddl.rs [hunk 3]
```diff
@@ -1814,18 +1814,45 @@ pub fn plan_create_materialized_view(
...
+            connection
+        }
+        mz_sql_parser::ast::KafkaConnection::Inline { .. } => unreachable!(),
```
Comment (guswynn) [inline src/sql/src/plan/statement/ddl.rs:1853]: helllll ya

File: test/kafka-sasl-plain/smoketest.td [hunk 4]
```diff
@@ -22,15 +22,18 @@ $ kafka-ingest format=avro topic=data schema=${schema} publish=true timestamp=1
 
 > CREATE SECRET sasl_password AS 'sekurity'
 
+# Ensure that connectors work with SSL basic_auth
```
Comment (guswynn) [inline test/kafka-sasl-plain/smoketest.td:25]: omg this file is so cleannnn now

**Review by guswynn — APPROVED** [review]
@sploiselle The only thing I don't feel qualified to review is the commented out test in `misc/dbt-materialize/tests/adapter/test_custom_materializations.py`, but its probably fine!

This is great! I'm sure the `SOURCE` tests you changed will make the future `SOURCE` version of this pr a bit easier!

File: src/sql-parser/src/ast/defs/ddl.rs (hunk 1 above)
Comment (sploiselle) [reply src/sql-parser/src/ast/defs/ddl.rs:959]: The `CONNECTION` gets printed by the connection object

Comment (guswynn) [reply src/sql-parser/src/ast/defs/ddl.rs:959]: oh nice!

File: src/sql/src/plan/statement/ddl.rs (hunk 2 above)
Comment (sploiselle) [reply src/sql/src/plan/statement/ddl.rs:1850]: I did not add a test for this, but it is the same condition as:
https://github.com/MaterializeInc/materialize/blob/main/test/kafka-ssl/smoketest.td#L207-L221

**Pull request merged**